server/server.go

```diff
@@ -132,12 +146,14 @@ func (c *Conn) write(v interface{}) error {
 
 // UpdateState sends registered users and open games to the client.
 func (c *Conn) UpdateState(id UserId) error {
-	if err := c.UserRegistered(users); err != nil {
-		log.Println(users, "UserRegistered")
+	state.Lock()
+	defer state.Unlock()
```

Comment:
Why do this vs putting `state.Unlock()` at the end of the function? It seems safer for future edits to the function so future you or me doesn't need to remember to keep moving it down. Are there other reasons or just that?

Reply:
It's actually safer to use a `defer`. What if we `panic()` between calling `state.Lock()` and `state.Unlock()`? Every call to [`sync/Mutex.Lock()`](https://golang.org/pkg/sync/#Mutex.Lock) simply waits if the mutex is already locked. If one goroutine fails without unlocking the mutex, it will remain locked forever. `defer`, on the other hand, lets us schedule unlocking for the end of the function. Deferred statements also run during panic recovery, which is the behavior we want.

Although Go doesn't use the term, Go panics are sort of like Ruby exceptions and Go defers are spiritually similar to Ruby try/catch blocks.